scale/node/resources/gpu_manager.py

```diff
@@ -0,0 +1,139 @@
+"""Defines the class that keeps track of GPU resources on a per node basis"""
+import logging
+
+logger = logging.getLogger(__name__)
+
+class GPUManager(object):
+    """The class that holds the GPU library and manages it"""
+    __GPUs = {}
+
+    @classmethod
+    def assign_gpus_for_job(cls, node_id, job_id, required_gpu_count):
+        """
+        method that assigns a specific GPU to a job id
+        :param node_id: the node ID
+        :param job_id: the job ID
+        :param required_gpu_count: required GPUs
+        """
+        logger.info("assigning GPUs, request came for node %s and job %s for %s GPUs", node_id, job_id, required_gpu_count)
+        assigned_gpu_count = 0
+        assignment_complete = False
+        for gpunum, gpustatus in cls.__GPUs[node_id].iteritems():
+            logger.info("entered loop looking for gpus to set, expecting to set %s GPUs", int(required_gpu_count))
+            if gpustatus == "reserved":
+                cls.__GPUs[node_id][gpunum] = job_id
+                logger.info("assigned %s to %s", gpunum, job_id)
+                assigned_gpu_count += 1
+            if assigned_gpu_count == int(required_gpu_count):
+                assignment_complete = True
+                break # assigned everything we need, exit loop
+        if assignment_complete:
+            return True
+        else:
+            return False #TODO revert assigned GPUs, try again?
```

Comment:
We can't split a job across several nodes, can we? If not and we aren't able to get the required number on a node, I think we need to revert assignments before moving to the next node.

Reply:
I agree. but should the gpu_manager assume the job will be cancelled and simply release the GPUs? or should that be handled by the create_job_exe_model method, or maybe even one level higher and deal with it in the manager?

Reply:
I think the gpu manager should release the gpus and return False and the scheduler should leave it in the queue like any other job with insufficient resources.